devdoc/waiter.md

```diff
@@ -0,0 +1,71 @@
+# `waiter` requirements
+
+## Overview
+
+`waiter` is a module that allows the user to move data asynchronously from producer to consumer without having to block the thread.
+
+Consumers call `waiter_register_notification` to register a `NOTIFICATION_CALLBACK` function to be called when data becomes available.
+
+Producers call `waiter_notify` to notify the waiter that data is available.
+
+`waiter_register_notification` and `waiter_notify` can be called from different threads in any order.
+
+## Exposed API
+```c
+#define WAITER_RESULT_VALUES \
+    WAITER_RESULT_OK, \
+    WAITER_RESULT_CANCELLED, \
+    WAITER_RESULT_ABANDONED
+
+MU_DEFINE_ENUM(WAITER_RESULT, WAITER_RESULT_VALUES);
+
+typedef struct WAITER_TAG* WAITER_HANDLE;
+typedef void(*NOTIFICATION_CALLBACK)(void* context, THANDLE(RC_PTR) data, WAITER_RESULT result);
+typedef void(*NOTIFY_COMPLETE_CALLBACK)(void* context, WAITER_RESULT result);
+
+    MOCKABLE_FUNCTION(, WAITER_HANDLE, waiter_create)
+    MOCKABLE_FUNCTION(, void, waiter_destroy, WAITER_HANDLE, waiter)
+    MOCKABLE_FUNCTION(, THANDLE(ASYNC_OP), waiter_register_notification, WAITER_HANDLE, waiter, NOTIFICATION_CALLBACK, notification_callback, void*, context);
+    MOCKABLE_FUNCTION(, THANDLE(ASYNC_OP), waiter_notify, WAITER_HANDLE, waiter, THANDLE(RC_PTR), data, NOTIFY_COMPLETE_CALLBACK, notify_complete_callback, void*, context);
+
+```
+
+### waiter_create
+```c
+    MOCKABLE_FUNCTION(, WAITER_HANDLE, waiter_create)
+```
+
+- `waiter_create` shall create the waiter and returns it.
+
+### waiter_destroy
+```c
+    MOCKABLE_FUNCTION(, void, waiter_destroy, WAITER_HANDLE, waiter)
+```
+
+- If `waiter_register_notification` has been called previously but `waiter_notify` has not been called, `waiter_destroy` shall call the `notification_callback` given to `waiter_register_notification` with `context` as the `context` given to `waiter_register_notification`, `data` as `NULL` and `result` as `WAITER_RESULT_ABANDONED`.
+- If `waiter_notify` has been called previously but `waiter_register_notification` has not been called, `waiter_destroy` shall call the `notify_complete_callback` given to `waiter_notify` with `context` as the `context` given to `waiter_notify`, and `result` as `WAITER_RESULT_ABANDONED`.
+- `waiter_destroy` shall destroy the given `waiter`.
+
+### waiter_register_notification
+```c
+    MOCKABLE_FUNCTION(, THANDLE(ASYNC_OP), waiter_register_notification, WAITER_HANDLE, waiter, NOTIFICATION_CALLBACK, notification_callback, void*, context);
+```
+- `waiter_register_notification` registers the given `notification_callback` to be called when `waiter_notify` is called.
+- If `waiter_notify` has been called previously, `waiter_register_notification` shall  call:
```

Comment:
>has been called previously [](http://example.com/codeflow?start=21&length=26)

so if data has been produced... and it is sort of in a waiting pattern for `waiter_register_notification` to be called... 

... at the time of `waiter_register_notification` call it looks like the callback is called from the same thread which called `waiter_register_notification`. 

And in that case, ASYNC_OP that is being returned is obsolete by the time when the function returns, as there's no cancellation possible.

Reply:
Correct, the operation complete synchronously and the ASYNC_OP that is returned is obsolete.

However if `waiter_notify` is called before `waiter_register_notification`, then the ASYNC_OP can be used to cancel the operation.